core/server/common/src/main/java/alluxio/master/journal/raft/RaftJournalSystem.java

```diff
@@ -327,6 +334,58 @@ public synchronized void losePrimacy() {
     LOG.info("Raft server successfully restarted");
   }
 
+  @Override
+  public synchronized void checkpoint() throws IOException {
+    mSnapshotAllowed.set(true);
+    CopycatClient client = createAndConnectClient();
+    try {
+      long start = System.currentTimeMillis();
+      LOG.info("Submitting empty journal entry to trigger snapshot");
+      // If snapshot requirements are fulfilled, a snapshot will be triggered
+      // after sending an empty journal entry to Copycat
+      CompletableFuture<Void> future = client.submit(new JournalEntryCommand(
+          JournalEntry.getDefaultInstance()));
+      try {
+        future.get(1, TimeUnit.MINUTES);
+      } catch (TimeoutException | ExecutionException e) {
+        LOG.info("Exception submitting entry to trigger snapshot: {}", e.toString());
+        throw new IOException("Exception submitting entry to trigger snapshot", e);
+      } catch (InterruptedException e) {
+        LOG.info("Interrupted when submitting entry to trigger snapshot: {}", e.toString());
+        Thread.currentThread().interrupt();
+        throw new CancelledException("Interrupted when submitting entry to trigger snapshot", e);
+      }
+      if (mStateMachine.getLastSnapshotStartTime() < start) {
```

Comment:
what if the snapshot is about to start, but just hasn't started yet? Perhaps we could retry for some period before giving up

Reply:
Add the retry check start time, thanks